vignettes_src/cuda.Rmd

```diff
@@ -0,0 +1,326 @@
+
+```cc
+template <>
+DEVICE sirs::real_t compare_device<sirs>(const dust::interleaved<sirs::real_t> state,
+                           const sirs::data_t& data,
+                           dust::interleaved<int> internal_int,
+                           dust::interleaved<sirs::real_t> internal_real,
+                           const int * shared_int,
+                           const sirs::real_t * shared_real,
+                           dust::rng_state_t<sirs::real_t>& rng_state) {
+  typedef sirs::real_t real_t;
+  const real_t exp_noise = shared_real[4];
+  const real_t incidence_modelled = state[3];
+  const real_t incidence_observed = data.incidence;
+  const real_t lambda = incidence_modelled +
+    dust::distr::rexp(rng_state, exp_noise);
+  return dust::dpois(incidence_observed, lambda, true);
+}
+```
+
+This is very similar to the CPU version
+
+```cc
+  real_t compare_data(const real_t * state, const data_t& data,
+                      dust::rng_state_t<real_t>& rng_state) {
+    const real_t incidence_modelled = state[3];
+    const real_t incidence_observed = data.incidence;
+    const real_t lambda = incidence_modelled +
+      dust::distr::rexp(rng_state, shared->exp_noise);
+    return dust::dpois(incidence_observed, lambda, true);
+  }
+```
+
+with similar differences to the update function:
+
+* argument types are different (interleaved types, internals and shared data passed explicitly)
+* Accessing of shared parameters is by position, not name
+* The `DEVICE` annotation
+
+## Developing a GPU model
+
+Debugging on a GPU is a pain, especially because there are typically many particles, and error recovery is not straightforward.  In addition, most continuous integration systems do not provide GPUs, so testing your GPU code becomes difficult.  To make this easier, `dust` allows running GPU code on the CPU - this will be typically slower than the CPU code, but allows easier debugging and verification that the model is behaving.  We use this extensively in `dust`'s tests and also in models built using `dust` that will run on the GPU.
+
+To do this, compile the model with your preferred real type, but set the `gpu` argument to `FALSE`
+
+```{r}
+sirs_cpu <- dust::dust(path, gpu = FALSE, real_t = "float")
+```
+
+Note that above the model is compiled with `g++`, not `nvcc`. However, the "device" code is still compiled into the model.  We can then initialise this with and without a `device_config` argument
+
+```{r}
+model1 <- sirs_cpu$new(list(), 0, 10, seed = 1L)
+model2 <- sirs_cpu$new(list(), 0, 10, device_config = 0L, seed = 1L)
+```
+
+And run the models using the "device" code and the normal CPU code:
+
+```{r}
+model1$run(10, device = FALSE)
+model2$run(10, device = TRUE)
+```
+
+These should be exactly the same, and this can form the basis of tests. Note that using `float` rather than `double` can throw up a few issues with algorithms, so it's worth checking with single-precision in your tests.
```

Comment:
May want to note you can run with `R -d cuda-gdb` to get a debugger, and perhaps also link your blog post about gdb + valgrind?